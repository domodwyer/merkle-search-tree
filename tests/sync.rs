use std::{
    collections::BTreeMap,
    fmt::{Display, Write},
    ops::RangeInclusive,
};

use merkle_search_tree::{
    diff::{diff, PageRange},
    MerkleSearchTree,
};

const N_VALUES: usize = 5_000;

#[derive(Debug, Default)]
struct SyncStats {
    /// Number of sync "rounds" required to converge the trees, where a "round"
    /// is a two way diff + key fetch.
    rounds: usize,

    keys_a_to_b: DirectionStats,
    keys_b_to_a: DirectionStats,
}

#[derive(Debug)]
struct DirectionStats {
    /// Minimum number of keys synced in a single round.
    min: usize,
    /// Maximum number of keys synced in a single round.
    max: usize,
    /// Total number of keys synced in one direction, across all syncs.
    total: usize,

    /// Number of rounds that sync'd 0 keys.
    zero_rounds: usize,
}

impl Default for DirectionStats {
    fn default() -> Self {
        Self {
            min: usize::MAX,
            max: usize::MIN,
            total: Default::default(),
            zero_rounds: Default::default(),
        }
    }
}

impl Display for DirectionStats {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{:>10} keys total\tmin {:>10}\tmax {:>10}\t{} zero rounds",
            self.total, self.min, self.max, self.zero_rounds
        )
    }
}

/// A test that drives convergence between two randomly generated reference
/// trees (random key/value pairs are deterministic across runs), and records
/// statistics describing the keys/rounds required to reach convergence.
///
/// MST/diff changes can change the efficiency of syncs, and this test surfaces
/// those changes.
#[test]
fn test_sync_rounds() {
    let mut out = String::new();
    let mut total_rounds = 0;
    let mut total_keys = 0;
    let mut n_rounds = 0;
    for i in 1..=30 {
        let v = do_sync(Lfsr(i));

        writeln!(
            &mut out,
            "[*] sync with seed {i} - total sync rounds: {}",
            v.rounds
        )
        .unwrap();
        writeln!(
            &mut out,
            "\ta->b: {}\tavg: {:<5} keys per round",
            v.keys_a_to_b,
            v.keys_a_to_b
                .total
                .checked_div(v.rounds)
                .unwrap_or_default(),
        )
        .unwrap();
        writeln!(
            &mut out,
            "\tb->a: {}\tavg: {:<5} keys per round",
            v.keys_b_to_a,
            v.keys_b_to_a
                .total
                .checked_div(v.rounds)
                .unwrap_or_default(),
        )
        .unwrap();
        writeln!(&mut out).unwrap();

        total_rounds += v.rounds;
        n_rounds += 1;
        total_keys += v.keys_a_to_b.total;
        total_keys += v.keys_b_to_a.total;
    }

    writeln!(&mut out).unwrap();
    writeln!(
        &mut out,
        "{total_rounds} total sync rounds needed to converge {n_rounds} tree \
		pairs (average {:.2} rounds, {} keys per pair)",
        total_rounds as f32 / n_rounds as f32,
        total_keys / n_rounds,
    )
    .unwrap();

    insta::assert_snapshot!(out);
}

fn do_sync(mut rand: Lfsr) -> SyncStats {
    let mut a = Node::default();
    let mut b = Node::default();

    // Populate the trees with disjoint keys
    for _ in 0..(N_VALUES / 2) {
        a.upsert(rand.next().to_string(), rand.next());
        b.upsert(rand.next().to_string(), rand.next());
    }

    // Populate the trees with identical key/value pairs
    for _ in 0..(N_VALUES / 2) {
        let key = rand.next().to_string();
        let value = rand.next();
        a.upsert(key.clone(), value);
        b.upsert(key.clone(), value);
    }

    let mut result = SyncStats::default();

    // Drive them to convergence, recording sync statistics.
    while a.1.root_hash() != b.1.root_hash() {
        result.rounds += 1;
        sync_round(&mut a, &mut b, &mut result.keys_a_to_b);
        sync_round(&mut b, &mut a, &mut result.keys_b_to_a);
    }

    result
}

/// Perform a single sync round, pulling differences from a into b.
fn sync_round(from: &mut Node, to: &mut Node, stats: &mut DirectionStats) {
    // First sync b from a, applying the "a is always right" merge rule.
    let a2 = from.clone();
    let a_tree = from.page_ranges();

    let mut to2 = to.clone();
    let want = diff(to2.page_ranges(), a_tree);

    let mut count = 0; // Number of keys upserted
    for range in want {
        for (k, v) in a2.key_range_iter(range.start().to_owned()..=range.end().to_owned()) {
            to.upsert(k.clone(), *v);
            count += 1;
        }
    }

    stats.min = stats.min.min(count);
    stats.max = stats.max.max(count);
    stats.total += count;

    if count == 0 {
        stats.zero_rounds += 1;
    }
}

/// A mock peer, storing `(key, value)` tuples and maintaining a
/// [`MerkleSearchTree`] of the store contents.
#[derive(Default, Clone)]
pub struct Node(BTreeMap<String, u16>, MerkleSearchTree<String, u16>);

impl Node {
    /// Store the given `key` & `value` in the node, replacing the previous
    /// value of `key`, if any.
    pub fn upsert(&mut self, key: String, value: u16) {
        self.1.upsert(key.clone(), &value);
        self.0.insert(key, value);
    }

    /// Return a serialised representation of the [`MerkleSearchTree`] pages for
    /// diff computations.
    pub fn page_ranges(&mut self) -> Vec<PageRange<'_, String>> {
        self.1.root_hash();
        self.1.serialise_page_ranges().unwrap()
    }

    /// Return an iterator over the specified inclusive range of keys.
    pub fn key_range_iter(
        &self,
        key_range: RangeInclusive<String>,
    ) -> impl Iterator<Item = (&String, &u16)> {
        self.0.range(key_range)
    }
}

/// Linear-feedback shift register based PRNG.
///
/// Generates 65,535 unique values before cycling.
#[derive(Debug)]
pub struct Lfsr(u16);

impl Default for Lfsr {
    fn default() -> Self {
        Self(42)
    }
}

impl Lfsr {
    #[allow(clippy::should_implement_trait)]
    pub fn next(&mut self) -> u16 {
        let lsb = self.0 & 1;
        self.0 >>= 1;
        if lsb == 1 {
            self.0 ^= 0xD008;
        }
        self.0
    }
}
