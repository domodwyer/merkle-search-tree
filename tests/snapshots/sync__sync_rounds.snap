---
source: tests/sync.rs
expression: out
---
[*] sync with seed 1 - total sync rounds: 2
	a->b:       5003 keys total	min          0	max       5003	1 zero rounds	avg: 2501  keys per round
	b->a:       7499 keys total	min          4	max       7495	0 zero rounds	avg: 3749  keys per round

[*] sync with seed 2 - total sync rounds: 3
	a->b:      11337 keys total	min          2	max       6333	0 zero rounds	avg: 3779  keys per round
	b->a:      12474 keys total	min          2	max       7500	0 zero rounds	avg: 4158  keys per round

[*] sync with seed 3 - total sync rounds: 4
	a->b:      18238 keys total	min          0	max       6903	1 zero rounds	avg: 4559  keys per round
	b->a:      18415 keys total	min         11	max       7458	0 zero rounds	avg: 4603  keys per round

[*] sync with seed 4 - total sync rounds: 2
	a->b:       7499 keys total	min          4	max       7495	0 zero rounds	avg: 3749  keys per round
	b->a:       5003 keys total	min          0	max       5003	1 zero rounds	avg: 2501  keys per round

[*] sync with seed 5 - total sync rounds: 3
	a->b:      11853 keys total	min          0	max       6852	1 zero rounds	avg: 3951  keys per round
	b->a:      13176 keys total	min          2	max       7500	0 zero rounds	avg: 4392  keys per round

[*] sync with seed 6 - total sync rounds: 2
	a->b:      11154 keys total	min       3658	max       7496	0 zero rounds	avg: 5577  keys per round
	b->a:       6803 keys total	min          0	max       6803	1 zero rounds	avg: 3401  keys per round

[*] sync with seed 7 - total sync rounds: 2
	a->b:      11146 keys total	min       3650	max       7496	0 zero rounds	avg: 5573  keys per round
	b->a:       6802 keys total	min          0	max       6802	1 zero rounds	avg: 3401  keys per round

[*] sync with seed 8 - total sync rounds: 4
	a->b:      12474 keys total	min          0	max       7500	1 zero rounds	avg: 3118  keys per round
	b->a:      11338 keys total	min          0	max       6334	1 zero rounds	avg: 2834  keys per round

[*] sync with seed 9 - total sync rounds: 4
	a->b:      18537 keys total	min          0	max       7473	1 zero rounds	avg: 4634  keys per round
	b->a:      18092 keys total	min          2	max       7500	0 zero rounds	avg: 4523  keys per round

[*] sync with seed 10 - total sync rounds: 4
	a->b:      18539 keys total	min          0	max       7474	1 zero rounds	avg: 4634  keys per round
	b->a:      18093 keys total	min          2	max       7500	0 zero rounds	avg: 4523  keys per round

[*] sync with seed 11 - total sync rounds: 3
	a->b:      18243 keys total	min       5000	max       7354	0 zero rounds	avg: 6081  keys per round
	b->a:      17104 keys total	min       2639	max       7500	0 zero rounds	avg: 5701  keys per round

[*] sync with seed 12 - total sync rounds: 4
	a->b:      18416 keys total	min         11	max       7458	0 zero rounds	avg: 4604  keys per round
	b->a:      18238 keys total	min          0	max       6903	1 zero rounds	avg: 4559  keys per round

[*] sync with seed 13 - total sync rounds: 3
	a->b:      14935 keys total	min          1	max       7492	0 zero rounds	avg: 4978  keys per round
	b->a:      19847 keys total	min       4880	max       7496	0 zero rounds	avg: 6615  keys per round

[*] sync with seed 14 - total sync rounds: 4
	a->b:      18429 keys total	min         11	max       7459	0 zero rounds	avg: 4607  keys per round
	b->a:      18235 keys total	min          0	max       6905	1 zero rounds	avg: 4558  keys per round

[*] sync with seed 15 - total sync rounds: 4
	a->b:      17302 keys total	min          0	max       7285	1 zero rounds	avg: 4325  keys per round
	b->a:      14393 keys total	min          2	max       7487	0 zero rounds	avg: 3598  keys per round

[*] sync with seed 16 - total sync rounds: 2
	a->b:       5003 keys total	min          0	max       5003	1 zero rounds	avg: 2501  keys per round
	b->a:       7499 keys total	min          4	max       7495	0 zero rounds	avg: 3749  keys per round

[*] sync with seed 17 - total sync rounds: 4
	a->b:      16214 keys total	min          0	max       7500	1 zero rounds	avg: 4053  keys per round
	b->a:      13218 keys total	min          0	max       7326	1 zero rounds	avg: 3304  keys per round

[*] sync with seed 18 - total sync rounds: 3
	a->b:      13179 keys total	min          3	max       7499	0 zero rounds	avg: 4393  keys per round
	b->a:      11854 keys total	min          0	max       6852	1 zero rounds	avg: 3951  keys per round

[*] sync with seed 19 - total sync rounds: 3
	a->b:      11083 keys total	min          0	max       6071	1 zero rounds	avg: 3694  keys per round
	b->a:      11561 keys total	min         30	max       7500	0 zero rounds	avg: 3853  keys per round

[*] sync with seed 20 - total sync rounds: 3
	a->b:      13175 keys total	min          3	max       7499	0 zero rounds	avg: 4391  keys per round
	b->a:      11853 keys total	min          0	max       6851	1 zero rounds	avg: 3951  keys per round

[*] sync with seed 21 - total sync rounds: 3
	a->b:      13176 keys total	min          3	max       7499	0 zero rounds	avg: 4392  keys per round
	b->a:      11848 keys total	min          0	max       6846	1 zero rounds	avg: 3949  keys per round

[*] sync with seed 22 - total sync rounds: 3
	a->b:       8229 keys total	min          2	max       7498	0 zero rounds	avg: 2743  keys per round
	b->a:      10245 keys total	min          0	max       5243	1 zero rounds	avg: 3415  keys per round

[*] sync with seed 23 - total sync rounds: 3
	a->b:      11902 keys total	min          0	max       6901	1 zero rounds	avg: 3967  keys per round
	b->a:      14383 keys total	min          1	max       7471	0 zero rounds	avg: 4794  keys per round

[*] sync with seed 24 - total sync rounds: 2
	a->b:       6803 keys total	min          0	max       6803	1 zero rounds	avg: 3401  keys per round
	b->a:      11155 keys total	min       3659	max       7496	0 zero rounds	avg: 5577  keys per round

[*] sync with seed 25 - total sync rounds: 3
	a->b:      12476 keys total	min          0	max       7475	1 zero rounds	avg: 4158  keys per round
	b->a:      14909 keys total	min          2	max       7500	0 zero rounds	avg: 4969  keys per round

[*] sync with seed 26 - total sync rounds: 4
	a->b:      16214 keys total	min          0	max       7500	1 zero rounds	avg: 4053  keys per round
	b->a:      13217 keys total	min          0	max       7326	1 zero rounds	avg: 3304  keys per round

[*] sync with seed 27 - total sync rounds: 2
	a->b:      12464 keys total	min       4964	max       7500	0 zero rounds	avg: 6232  keys per round
	b->a:       7483 keys total	min          0	max       7483	1 zero rounds	avg: 3741  keys per round

[*] sync with seed 28 - total sync rounds: 2
	a->b:       6802 keys total	min          0	max       6802	1 zero rounds	avg: 3401  keys per round
	b->a:      11146 keys total	min       3650	max       7496	0 zero rounds	avg: 5573  keys per round

[*] sync with seed 29 - total sync rounds: 2
	a->b:      12248 keys total	min       5000	max       7248	0 zero rounds	avg: 6124  keys per round
	b->a:      14744 keys total	min       7249	max       7495	0 zero rounds	avg: 7372  keys per round

[*] sync with seed 30 - total sync rounds: 2
	a->b:      12464 keys total	min       4964	max       7500	0 zero rounds	avg: 6232  keys per round
	b->a:       7483 keys total	min          0	max       7483	1 zero rounds	avg: 3741  keys per round


89 total sync rounds needed to converge 30 tree pairs (average 2.97 rounds, 25421 keys per pair)

